common/src/main/java/io/netty/util/internal/ThreadCleaner.java

```diff
@@ -0,0 +1,132 @@
+import java.lang.ref.WeakReference;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Allows a way to register some {@link Runnable} that will executed once there are no references to the {@link Thread}
+ * anymore. This typically happens once the {@link Thread} dies / completes.
+ */
+public final class ThreadCleaner {
+
+    // This will hold a reference to the ThreadCleanerReference which will be removed once we called cleanup()
+    private static final Set<ThreadCleanerReference> LIVE_SET = new ConcurrentSet<ThreadCleanerReference>();
+    private static final ReferenceQueue<Object> REFERENCE_QUEUE = new ReferenceQueue<Object>();
+    private static final AtomicBoolean CLEANER_RUNNING = new AtomicBoolean(false);
+
+    private static final Runnable CLEANER_TASK = new Runnable() {
+        @Override
+        public void run() {
+            boolean interrupted = false;
+            for (;;) {
+                // Keep on processing as long as the LIVE_SET is not empty and once it becomes empty
+                // See if we can let this thread complete.
+                while (!LIVE_SET.isEmpty()) {
+                    try {
+                        ThreadCleanerReference reference =
+                                (ThreadCleanerReference) REFERENCE_QUEUE.remove();
+                        try {
+                            reference.cleanup();
+                        } finally {
+                            LIVE_SET.remove(reference);
+                        }
+                    } catch (InterruptedException ex) {
+                        // Just consume and move on
+                        interrupted = true;
+                    }
+                }
+                CLEANER_RUNNING.set(false);
+
+                // Its important to first access the LIVE_SET and then CLEANER_RUNNING to ensure correct
+                // behavior in multi-threaded environments.
+                if (LIVE_SET.isEmpty() || !CLEANER_RUNNING.compareAndSet(false, true)) {
+                    // There was nothing added after we set STARTED to false or some other cleanup Thread
+                    // was started already so its safe to let this Thread complete now.
+                    break;
+                }
+            }
+            if (interrupted) {
+                // As we catched the InterruptedException above we should mark the Thread as interrupted.
+                Thread.currentThread().interrupt();
+            }
+        }
+    };
+
+    /**
+     * Register the given {@link Thread} for which the {@link Runnable} will be executed once there are no references
+     * to the object anymore, which typically happens once the {@link Thread} dies.
+     *
+     * This should only be used if there are no other ways to execute some cleanup once the {@link Thread} dies as
+     * it is not a cheap way to handle the cleanup.
+     */
+    public static void register(Thread thread, Runnable cleanupTask) {
```

Comment:
this class is `internal` ... consider:
1. exposing a method where the user can directly create a `ThreadCleanerReference` to avoid allocating a `Runnable` and `ThreadCleanerReference`
2. Make the class package private and you can use the same technique as above to reduce object allocation.

Reply:
I am not concerned about this as this should be something that is done very frequently and I think this design is cleaner.

Reply:
Agreed the current APIs are nicer. However it is internal and doesn't need to be generic right? We could potentially save 1 object allocation per thread local. Since it is internal we can always change it later but maybe nice to keep thread local overhead down if we don't have to be generic.

Reply:
@Scottmitch as discussed offline I will merge it as it is for now and we can address this if we find out its an issue later on. The current design is nicer :)

Reply:
sgtm ... internal == change later if necessary

Reply:
exactly 👍